artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/SCRAMPropertiesLoginModule.java

```diff
@@ -0,0 +1,202 @@
+import org.apache.activemq.artemis.utils.PasswordMaskingUtil;
+
+/**
+ * Login modules that uses properties files similar to the {@link PropertiesLoginModule}. It can
+ * either store the username-password in plain text or in an encrypted/hashed form. the
+ * {@link #main(String[])} method provides a way to prepare unencrypted data to be encrypted/hashed.
+ */
+public class SCRAMPropertiesLoginModule extends PropertiesLoader implements AuditLoginModule {
+
+   private static final String SEPARATOR = ":";
+   private static final int MIN_ITERATIONS = 4096;
+   private Subject subject;
+   private CallbackHandler callbackHandler;
+   private Properties users;
+   private Map<String, Set<String>> roles;
+   private UserData userData;
+   private String user;
+   private final Set<Principal> principals = new HashSet<>();
+
+   @Override
+   public void initialize(Subject subject, CallbackHandler callbackHandler, Map<String, ?> sharedState,
+                          Map<String, ?> options) {
+      this.subject = subject;
+      this.callbackHandler = callbackHandler;
+
+      init(options);
+      users = load(PropertiesLoginModule.USER_FILE_PROP_NAME, "user", options).getProps();
+      roles = load(PropertiesLoginModule.ROLE_FILE_PROP_NAME, "role", options).invertedPropertiesValuesMap();
+
+   }
+
+   @Override
+   public boolean login() throws LoginException {
+      NameCallback nameCallback = new NameCallback("Username: ");
+      executeCallbacks(new Callback[] {nameCallback});
+      user = nameCallback.getName();
+      if (user == null) {
+         throw new FailedLoginException("User is null");
+      }
+      String password = users.getProperty(user);
+      if (password == null) {
+         throw new FailedLoginException("User does not exist: " + user);
+      }
+      if (PasswordMaskingUtil.isEncMasked(password)) {
+         String[] unwrap = PasswordMaskingUtil.unwrap(password).split(SEPARATOR);
+         userData = new UserData(unwrap[0], Integer.parseInt(unwrap[1]), unwrap[2], unwrap[3]);
+      } else {
+         DigestCallback digestCallback = new DigestCallback();
+         HmacCallback hmacCallback = new HmacCallback();
+         executeCallbacks(new Callback[] {digestCallback, hmacCallback});
+         byte[] salt = generateSalt();
+         try {
+            ScramUtils.NewPasswordStringData data =
+                     ScramUtils.byteArrayToStringData(ScramUtils.newPassword(password, salt, 4096,
+                                                                             digestCallback.getDigest(),
+                                                                             hmacCallback.getHmac()));
+            userData = new UserData(data.salt, data.iterations, data.serverKey, data.storedKey);
+         } catch (ScramException e) {
+            throw new LoginException();
+         }
+      }
+      return true;
+   }
+
+   private static byte[] generateSalt() {
+      byte[] salt = new byte[24];
```

Comment:
Any reason to use 24 specifically rather than say 32?

Reply:
No specific reason, the SASL-SCRAM-FOR-JAVA uses the same size so I used it here. Any suggestions or concerns for an alternative size? I checked the spec and it does not gives any advice or recommendation just that a salt should be used:

> Salt: A random octet string that is combined with a password
>       before applying a one-way encryption function.  This value is used
>       to protect passwords that are stored in an authentication
>       database.

Reply:
I mainly just thought 24 seemed unusual is all. I belive Qpid Broker-J uses 32, I'd perhaps do the same as as nice power of 2.